ocrd_network/ocrd_network/deployer.py

```diff
@@ -0,0 +1,302 @@
+Each Processing Host may have several Processing Workers.
+Each Processing Worker is an instance of an OCR-D processor.
+"""
+
+from __future__ import annotations
+from typing import Dict, Union, Optional
+from paramiko import SSHClient
+from pathlib import Path
+from re import search as re_search
+from time import sleep
+
+
+from ocrd_utils import getLogger
+from ocrd_network.deployment_config import *
+from ocrd_network.deployment_utils import (
+    create_docker_client,
+    create_ssh_client,
+    CustomDockerClient,
+    DeployType,
+    HostData,
+)
+from ocrd_network.rabbitmq_utils import RMQPublisher
+
+
+class Deployer:
+    """ Class to wrap the deployment-functionality of the OCR-D Processing-Servers
+
+    Deployer is the one acting. Config is for representation of the config-file only. DeployHost is
+    for managing information, not for actually doing things.
+    """
+
+    def __init__(self, config: ProcessingServerConfig) -> None:
+        """
+        Args:
+            config: Parsed processing-server-configuration
+        """
+        self.log = getLogger(__name__)
+        self.config = config
+        self.hosts = HostData.from_config(config.hosts)
+        self.mongo_pid = None
+        self.mq_pid = None
+
+    def kill_all(self) -> None:
+        """ kill all started services: workers, database, queue
+
+        The order of killing is important to optimize graceful shutdown in the future. If RabbitMQ
+        server is killed before killing Processing Workers, that may have bad outcome and leave
+        Processing Workers in an unpredictable state
+        """
+        self.kill_hosts()
+        self.kill_mongodb()
+        self.kill_rabbitmq()
+
+    def deploy_hosts(self, rabbitmq_url: str, mongodb_url: str) -> None:
+        for host in self.hosts:
+            self.log.debug(f'Deploying processing workers on host: {host.config.address}')
+
+            if (any(p.deploy_type == DeployType.native for p in host.config.processors)
+                    and not host.ssh_client):
+                host.ssh_client = create_ssh_client(
+                    host.config.address,
+                    host.config.username,
+                    host.config.password,
+                    host.config.keypath
+                )
+            if (any(p.deploy_type == DeployType.docker for p in host.config.processors)
+                    and not host.docker_client):
+                host.docker_client = create_docker_client(
+                    host.config.address,
+                    host.config.username,
+                    host.config.password,
+                    host.config.keypath
+                )
+
+            for processor in host.config.processors:
+                self._deploy_processing_worker(processor, host, rabbitmq_url, mongodb_url)
+
+            if host.ssh_client:
+                host.ssh_client.close()
+            if host.docker_client:
+                host.docker_client.close()
```

Comment:
Why close here? These could be used for stopping. (In the Docker case, does it even work to recreate a CustomDockerClient via `create_docker_client` in `kill_hosts`?)

Reply:
Right. The way it is implemented currently is to open/close connections when creating the processing workers and then again when killing processing workers. It's not efficient that way (since more connections are opened and closed) but at least we don't have to keep track of the connection state and restore it in case of connection fails.

Note that in the processing server and processing workers we preserve the connection state to the RabbitMQ server but we still don't have a reconnection mechanism (it's there just as a concept in the `ocrd-webapi-implementation` repo).

Reply:
I didn't (and still don't) see the benefit of keeping the connection open. Connection creating seems fast (and therefore cheap) to me so I thought keeping a connection for an unknown amount of time around does not make sense. Also this way there is no need to worry about whether the connection dies or something else happens to it.

It works btw to recreate a DockerClient, this is the way it is done for killing the mongodb and rabbitmq container at a shutdown request.

Reply:
> It works btw to recreate a DockerClient, this is the way it is done for killing the mongodb and rabbitmq container at a shutdown request.

Ok, I can see why it works now. You build upon the Docker daemon's internal ID mechanism for containers, which docker-py gives access to. 

So for MongoDB, RabbitMQ and `start_docker_processor`, I agree it's better to re-establish a connection when you want to stop the running container. But for `start_native_processor` we don't have this kind of manager in the background. Querying the PID and then killing it later is fragile IMO. (What if the process is already gone, or forked and detached?) I understand that [keeping the channel open](https://github.com/OCR-D/core/pull/974#discussion_r1121549686) might also be fragile, though. Perhaps experience will teach us best, so let's keep it (maybe with some comment).

Reply:
Could we reference this discussion in a short comment if this does become a problem in the future?

Reply:
Probably we should.

Reply:
continued in https://github.com/OCR-D/core/issues/1037